### PR TITLE
Refactored QueryPlanVisitor with new OpenLineageContext type

### PR DESCRIPTION
**Pull request by collado-mike**
### Problem
Part of the implementation of https://github.com/OpenLineage/OpenLineage/pull/437 - this is a large PR, but is entirely refactoring existing code. The principal changes are

1. Creates the `OpenLineageContext` class that holds references to other visitors, the (optional) SparkSession, SparkContext, and the OpenLineage client
2. Creates the `DatasetFactory` and moves utility methods from `PlanUtils` there in order to construct input or output datasets
3. Moving the QueryPlanVisitor to the `io.openlineage.spark.api` package
4. Add a non-null, final reference to `OpenLineageContext` to the `QueryPlanVisitor` constructor
5. Update all visitor code to reference the `DatasetFactory` or utility methods for Dataset construction

The first two items are posted as separate commits. The last three items couldn't be reasonably split up and the vast majority of the changes are outcomes of the last point.

The primary goal for this change is to ready the `QueryPlanVisitor` API for extension by code outside of this library. Extensions need a way to delegate to visitors outside of their own or this central codebase (e.g., a visitor for a command that wraps a `LogicalRelation` may not know how to deal with every type of `LogicalRelation` it may wrap). I also wanted to avoid the input/output Dataset wrapper delegates we have been using. We also need a common way to inject `SparkSession`s and other parameters, which we've done manually as we've written new visitors. 

For review, I'd look at the first two commits to see what's contained in the new API code, then review the third commit for structure, not content (the logic hasn't been changed anywhere - these are all changes to make the compiler happy).

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [ ] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [ ] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [ ] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)

### CODE REVIEW

**Review by pawel-big-lebowski — APPROVED** [review]
The change is really necessary and the sooner it gets merged the better as it affects all the other PRs. I put some minor comments for consideration.

**Review by mobuchowski — APPROVED** [review]

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/ContextFactory.java [hunk 1]
```diff
@@ -19,41 +20,47 @@
...
   public final EventEmitter sparkContext;
 
   public void close() {
     sparkContext.close();
   }
 
   public RddExecutionContext createRddExecutionContext(int jobId) {
-    return new RddExecutionContext(jobId, sparkContext);
+    OpenLineageContext olContext =
+        new OpenLineageContext(
```
[the remark this answers is not in the document]
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/ContextFactory.java:32]: Good point re: the builder proposal. I think I'll use the Lombok annotation here.

Re: the `SparkSession.getOrCreate`, I'm actually hesitant to create a new `SparkSession` if one has been closed. If there's no thread-local open session, it'll create a global one, which can interfere with the application's expected state.

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AlterTableAddColumnsCommandVisitor.java [hunk 2]
```diff
@@ -2,43 +2,54 @@
...
   @Override
-  @SneakyThrows
-  public List<OpenLineage.Dataset> apply(LogicalPlan x) {
-    CatalogTable catalogTable =
-        sparkSession
-            .sessionState()
-            .catalog()
-            .getTableMetadata(((AlterTableAddColumnsCommand) x).table());
+  public List<OpenLineage.OutputDataset> apply(LogicalPlan x) {
+    Optional<CatalogTable> tableOption =
```
[the remark this answers is not in the document]
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AlterTableAddColumnsCommandVisitor.java:26]: Good call. I'll add this

File: integration/spark/src/main/common/java/io/openlineage/spark/api/OpenLineageContext.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan;
+import org.apache.spark.sql.execution.QueryExecution;
+import scala.PartialFunction;
+
+/**
+ * Context holder with references to several required objects during construction of an OpenLineage
+ * {@link io.openlineage.client.OpenLineage.RunEvent}.
+ */
+@Value
+public class OpenLineageContext {
```
[the remark this answers is not in the document]
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/api/OpenLineageContext.java:33]: I think we should consider this for a future PR

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/ContextFactory.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/ContextFactory.java]: Take a look at the builder, specifically at the comments at https://github.com/OpenLineage/OpenLineage/pull/447/files#diff-1a7751b29663d3eff18c61b034dae4dda1e2443b4e3acbd22f6a8eded5e21e80R70-R78 . Let me know if this makes sense to you.

[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/ContextFactory.java]: :thumbsup:

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java [hunk 4]
```diff
@@ -67,7 +68,7 @@
   private List<URI> outputs;
   private String jobSuffix;
 
-  public RddExecutionContext(int jobId, EventEmitter sparkContext) {
+  public RddExecutionContext(OpenLineageContext context, int jobId, EventEmitter sparkContext) {
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java:71]: _Minor_: Looks like `sparkContext` should be renamed to `eventEmitter`?

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AlterTableRenameCommandVisitor.java [hunk 5]
```diff
@@ -1,46 +1,38 @@
...
 @Slf4j
 public class AlterTableRenameCommandVisitor
-    extends QueryPlanVisitor<AlterTableRenameCommand, OpenLineage.Dataset> {
-
-  private final SparkSession sparkSession;
+    extends QueryPlanVisitor<AlterTableRenameCommand, OpenLineage.OutputDataset> {
 
-  public AlterTableRenameCommandVisitor(SparkSession sparkSession) {
-    this.sparkSession = sparkSession;
+  public AlterTableRenameCommandVisitor(OpenLineageContext openLineage) {
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/AlterTableRenameCommandVisitor.java:24]: _Minor_: I'd try and keep naming consistent. Looks like `context` is the winner?

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/BigQueryNodeVisitor.java [hunk 6]
```diff
@@ -22,12 +24,14 @@
  * bigquery://&lt;projectId&gt;.&lt;.datasetId&gt;.&lt;tableName&gt;</code> . The namespace for
  * bigquery tables is always <code>bigquery</code> and the name is the FQN.
  */
-public class BigQueryNodeVisitor extends QueryPlanVisitor<LogicalPlan, OpenLineage.Dataset> {
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/plan/BigQueryNodeVisitor.java:25]: _Minor_: You may want to update the class javadocs as the class definition has change? That is, `D` can either be an input or output dataset.

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/util/JdbcUtils.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+public class JdbcUtils {
+  /**
+   * JdbcUrl can contain username and password this method clean-up credentials from jdbcUrl and
+   * strip the jdbc prefix from the url
+   */
+  public static String sanitizeJdbcUrl(String jdbcUrl) {
+    jdbcUrl = jdbcUrl.substring(5);
+    return jdbcUrl
+        .replaceAll(PlanUtils.SLASH_DELIMITER_USER_PASSWORD_REGEX, "@")
```
[the remark this answers is not in the document]
Comment (wslulciuc) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/util/JdbcUtils.java:11]: +1

File: integration/spark/src/main/common/java/io/openlineage/spark/api/DatasetFactory.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
+package io.openlineage.spark.api;
+
+import io.openlineage.client.OpenLineage;
+import io.openlineage.spark.agent.util.DatasetIdentifier;
+import io.openlineage.spark.agent.util.PlanUtils;
+import java.net.URI;
+import org.apache.spark.sql.types.StructType;
+
+/** @param <D> */
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/DatasetFactory.java:9]: _Minor_: I would add a description for `D` and some javadocs for the class ;)

File: integration/spark/src/main/common/java/io/openlineage/spark/api/DatasetFactory.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+   */
+  private static OpenLineage.DatasetFacets datasetFacet(
+      OpenLineage openLineage, StructType schema, String namespaceUri) {
+    return openLineage
+        .newDatasetFacetsBuilder()
+        .schema(PlanUtils.schemaFacet(openLineage, schema))
+        .dataSource(PlanUtils.datasourceFacet(openLineage, namespaceUri))
+        .build();
+  }
+}
```
Comment (wslulciuc) [inline integration/spark/src/main/common/java/io/openlineage/spark/api/DatasetFactory.java:109]: Nicely defined set of factory methods 👌

**Review by wslulciuc — APPROVED** [review]
Left some minor comments, but AMAZING work on adding both the `OpenLineageContext` and `DatasetFactory` classes to improve overall usability 💯 🥇

File: integration/spark/src/main/common/java/io/openlineage/spark/api/OpenLineageContext.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/src/main/common/java/io/openlineage/spark/api/OpenLineageContext.java:34]: A good example for this is `alter table` v2 command. When implementing a visitor, we have available old schema and list of changes to apply. We can also extract current schema of a table (which is good for `complete` event) but we don't know within a visitor whether `START` or `COMPLETE` event is generated. The behavior of getting the actual schema may differ for `START` events, as generating event and modifying table happens at the same time. 

I agree this should not block this PR, let's merge it.

File: integration/spark/src/main/common/java/io/openlineage/spark/api/DatasetFactory.java (hunk 8 above)
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/api/DatasetFactory.java:9]: I swear I wrote these javadocs, but... they vanished... 😅

File: integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java (hunk 4 above)
Comment (collado-mike) [reply integration/spark/src/main/common/java/io/openlineage/spark/agent/lifecycle/RddExecutionContext.java:71]: I've done this in one of the other PRs- I'd rather not go through the hassle of fixing the conflicts here :)

**Pull request merged**